README.md
silc-toolkit
============
Version 1.1.10


ChangeLog
commit 8cb801cf6482666818e721822ce81c81ec818908
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Fri Aug 7 14:48:46 2009 +0300

    More string format fixes in silcd and client libary

commit 97fa30748eed3f4248bd213ae3ab9e742fd36c71
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Sun Aug 2 10:11:20 2009 +0300

    configure: changed AC_PROG_LIBTOOL order to fix disabling shared libs

commit 35b767c59f0741fc58f0c4d07ccd8d2ce4134a35
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Fri Jul 31 22:38:06 2009 +0300

    configure: check threads support in OpenBSD.

commit 1598b3a51b51a434037461ccd35487bc0df3137c
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Fri Jul 31 22:32:57 2009 +0300

    Fixed string format vulnerability in client entry handling.
    
    Reported and patch provided by William Cummings.

commit cb95ad2d35f1796567fbebc9e848c10d4bd2e7b9
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Mon Jun 8 08:58:35 2009 +0300

    silcd: Fixed IDENTIFY command reply handling for channels
    
    The IDENTIFY comamnd reply for channels were broken because the channel ID
    parsing always failed.  This would cause issues like USERS command not
    returning the users for a channel, making the channel appear empty, in some
    specific cases.
    
    Fixed also the USERS command reply to not query the channel information
    from router more than once.

commit 59f343801620f8b02933d419efb111d5928b11b8
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Mon Jun 8 08:03:02 2009 +0300

    silcd: Abort active protocols in DETACH command
    
    When detaching client, which results into closing of the actual connection
    it is important to abort any active protocols, such as rekeys, because
    they cannot be completed with detached clients.

commit aef788f086376673eef0939aa3ef74810318d5f3
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Sun Jun 7 20:04:16 2009 +0300

    silcd: Added SKR to debug levels

commit 3749c6ee56f1df40934bab1aae6baeb334206b69
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Tue Jun 2 08:01:25 2009 +0300

    silcd: Save debug message in ring buffer
    
    In debug version save 3000 last debug messages in ring buffer so that they
    are available in core files.

commit 7a10e386467421bfac0ba425289250e61e566ce0
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Mon May 25 20:44:07 2009 +0300

    configure.ad: Fixed configure to support new autoconf

commit 7d0cb0875ea6adaac843bc8d7ee1d74666c99afd
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Mon May 25 20:43:22 2009 +0300

    silcd: fixed Deny hostname comparison
    
    Use the silc_strin_match instead of silc_string_compare to match
    hostnames and IP addresses in silcd.conf.  Fixed for Deny but changed
    in all the other places as well.

commit a576439040a03e4b98b96b5834e504da00ced6e5
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Wed Mar 11 17:04:20 2009 +0200

    Added proper copyright notices to COPYING and BSD files

commit ae6052d215699d28c10af64da8066b7b5175f58d
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Tue Mar 3 08:02:27 2009 +0200

    silcd: IDENTIFY command reply didn't save userinfo correctly
    
    The IDENTIFY command reply didn't duplicate the userinfo when adding
    new client thus causing weird crashes later when freeing the userinfo.

commit 2e7268cde1710b4901e6db52c0868bd597b88dbc
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Tue Mar 3 07:52:58 2009 +0200

    silcd: IDENTIFY command reply didn't save client correctly
    
    The IDENTIFY command reply added sometimes new client without
    duplicating the nickname of the client thus causing odd crashes later
    when trying to free the nickname.

commit 226f97eb5ab714b58bfb2cba0cb50d025884bb13
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Fri Feb 27 16:22:36 2009 +0200

    PacketEngine: Don't free underlaying stream in packet stream create error
    
    If packet stream creation failed it freed the stream given as argument.
    This is wrong.  It is the caller's responsibility to free it if the
    packet stream creation failed.

commit f17f0003337914b58ced7b318ce8455ffc2a9e51
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Fri Feb 27 16:16:11 2009 +0200

    silcd: handle SIGNOFF notify for local clients correctly
    
    In some rare cases router can send SIGNOFF notify to server for a client
    that is locally connected to the server.  The server didn't close the
    client's connection but freed the entry.  This later caused various
    different crashes when accessing the freed the client.

commit 888990de6ce49969c38b46385644f946dabab5d1
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Thu Feb 26 08:07:56 2009 +0200

    Fixed unsigned <-> signed char issues

commit dc0e9d3b61c1adab40d23e48486f857f14bdcbd8
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Thu Feb 26 08:06:02 2009 +0200

    SKE: When failure is received mark SKE always failed
    
    Mark the SKE failed even if we don't receive error from remote.  Fixes
    crash where the callback is called back to application without valid
    key material and without error status.

commit 4993b03b9ccd85373a60f9cf5c8d99a865898fda
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Sat Jan 17 11:42:07 2009 +0200

    silcd: increment stat.clients statistics whenever adding new client
    
    When new client is added to server the stat.clients must always be
    incremented.  Otherwise it is possible to go under zero when removing
    the clients by server.

commit 454d63a3ea0c954f9d40b9e7725fd9ea48503f4c
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Thu Jan 15 12:21:23 2009 +0200

    silcd: do not create dynamic connections on non-dynamic server

commit c0a3e19503062eb3aab8335d97e27e348595b81d
Author: Pekka Riikonen <priikone@silcnet.org>
Date:   Mon Jan 12 09:03:41 2009 +0200

    Added scripts/release

commit 9504fdcc21339629d03619684f3ffdc0601e0a7e

lib/silccrypt/rsa.c
/*

  rsa.c 	RSA Public and Private key generation functions,
 	   	RSA encrypt and decrypt functions.

  Author: Pekka Riikonen <priikone@silcnet.org>

  Copyright (C) 1997 - 2007 Pekka Riikonen

  The contents of this file are subject to one of the Licenses specified 
  in the COPYING file;  You may not use this file except in compliance 
  with the License.

  The software distributed under the License is distributed on an "AS IS"
  basis, in the hope that it will be useful, but WITHOUT WARRANTY OF ANY
  KIND, either expressed or implied.  See the COPYING file for more
  information.

  Created: Sat Mar  1 13:26:45 1997 pekka

  RSA public key cryptographic algorithm used in this distribution is:

  	Key generation:
  	p, q		primes
  	p != q
 	n = p * q	modulus

 	Public key exponent:
 	e   relatively prime to (p-1) * (q-1)
 	Private key exponent:
 	d = e ^ -1 mod lcm(((p-1) * (q-1)))

 	Encryption:
 	c = m ^ e mod n
 	Decryption:
 	m = c ^ d mod n

  Supports CRT (Chinese Remainder Theorem) for private key operations.

  The SSH's (Secure Shell), PGP's (Pretty Good Privacy) and RSAREF
  Toolkit were used as reference when coding this implementation. They
  all were a big help for me.

  I also suggest reading Bruce Schneier's; Applied Cryptography, Second
  Edition, John Wiley & Sons, Inc. 1996. This book deals about RSA and
  everything else too about cryptography.

*/
/* $Id$ */

/*
   ChangeLog

   o Mon Feb 12 11:20:32 EET 2001  Pekka

     Changed RSA private exponent generation to what PKCS #1 suggests.  We
     try to find the smallest possible d by doing modinv(e, lcm(phi)) instead
     of modinv(e, phi).  Note: this is not security fix but optimization.

   o Tue Feb 20 13:58:58 EET 2001  Pekka

     Set key->bits in rsa_generate_key.  It is the modulus length in bits.
     The `tmplen' in encrypt, decrypt, sign and verify PKCS API functions
     is now calculated by (key->bits + 7) / 8.  It is the length of one block.

   o Sat Mar 16 18:27:19 EET 2002  Pekka

     Use the SilcRng sent as argument to SILC_PKCS_API_INIT in prime
     generation.

   o Sat Sep 26 19:59:48 EEST 2002  Pekka

     Fixed double free in public key setting.  Use a bit larger e as
     starting point in key generation.
*/

#include "silc.h"
#include "rsa.h"

/* Generates RSA public and private keys. Primes p and q that are used
   to compute the modulus n has to be generated before calling this. They
   are then sent as argument for the function. */

SilcBool silc_rsa_generate_keys(SilcUInt32 bits, SilcMPInt *p, SilcMPInt *q,
				void **ret_public_key, void **ret_private_key)
{
  RsaPublicKey *pubkey;
  RsaPrivateKey *privkey;
  SilcMPInt phi, hlp;
  SilcMPInt div, lcm;
  SilcMPInt pm1, qm1;

  *ret_public_key = pubkey = silc_calloc(1, sizeof(*pubkey));
  if (!pubkey)
    return FALSE;

  *ret_private_key = privkey = silc_calloc(1, sizeof(*privkey));
  if (!privkey)
    return FALSE;

  /* Initialize variables */
  silc_mp_init(&privkey->n);
  silc_mp_init(&privkey->e);
  silc_mp_init(&privkey->d);
  silc_mp_init(&privkey->dP);
  silc_mp_init(&privkey->dQ);
  silc_mp_init(&privkey->qP);
  silc_mp_init(&phi);
  silc_mp_init(&hlp);
  silc_mp_init(&div);
  silc_mp_init(&lcm);
  silc_mp_init(&pm1);
  silc_mp_init(&qm1);

  /* Set modulus length */
  privkey->bits = bits;

  /* Compute modulus, n = p * q */
  silc_mp_mul(&privkey->n, p, q);

  /* phi = (p - 1) * (q - 1) */
  silc_mp_sub_ui(&pm1, p, 1);
  silc_mp_sub_ui(&qm1, q, 1);
  silc_mp_mul(&phi, &pm1, &qm1);

  /* Set e, the public exponent. We try to use same public exponent
     for all keys. Also, to make encryption faster we use small
     number. */
  silc_mp_set_ui(&privkey->e, 65533);
 retry_e:
  /* See if e is relatively prime to phi. gcd == greates common divisor,
     if gcd equals 1 they are relatively prime. */
  silc_mp_gcd(&hlp, &privkey->e, &phi);
  if ((silc_mp_cmp_ui(&hlp, 1)) > 0) {
    silc_mp_add_ui(&privkey->e, &privkey->e, 2);
    goto retry_e;
  }

  /* Find d, the private exponent, e ^ -1 mod lcm(phi). */
  silc_mp_gcd(&div, &pm1, &qm1);
  silc_mp_div(&lcm, &phi, &div);
  silc_mp_modinv(&privkey->d, &privkey->e, &lcm);

  /* Optimize d with CRT. */
  silc_mp_mod(&privkey->dP, &privkey->d, &pm1);
  silc_mp_mod(&privkey->dQ, &privkey->d, &qm1);
  silc_mp_modinv(&privkey->qP, q, p);
  silc_mp_set(&privkey->p, p);
  silc_mp_set(&privkey->q, q);

  silc_mp_uninit(&phi);
  silc_mp_uninit(&hlp);
  silc_mp_uninit(&div);
  silc_mp_uninit(&lcm);
  silc_mp_uninit(&pm1);
  silc_mp_uninit(&qm1);

  /* Set public key */
  silc_mp_init(&pubkey->n);
  silc_mp_init(&pubkey->e);
  pubkey->bits = privkey->bits;
  silc_mp_set(&pubkey->n, &privkey->n);
  silc_mp_set(&pubkey->e, &privkey->e);

  return TRUE;
}

/* RSA public key operation */

SilcBool silc_rsa_public_operation(RsaPublicKey *key, SilcMPInt *src,
				   SilcMPInt *dst)
{
  /* dst = src ^ e mod n */
  silc_mp_pow_mod(dst, src, &key->e, &key->n);
  return TRUE;
}

/* RSA private key operation */

SilcBool silc_rsa_private_operation(RsaPrivateKey *key, SilcMPInt *src,
				    SilcMPInt *dst)
{
  SilcMPInt tmp;

  silc_mp_init(&tmp);

  /* dst = (src ^ dP mod p) */
  silc_mp_pow_mod(dst, src, &key->dP, &key->p);

  /* tmp = (src ^ dQ mod q) */
  silc_mp_pow_mod(&tmp, src, &key->dQ, &key->q);

  /* dst = (dst - tmp) * qP mod p */
  silc_mp_sub(dst, dst, &tmp);
  silc_mp_mul(dst, dst, &key->qP);
  silc_mp_mod(dst, dst, &key->p);

  /* dst = (q * dst) + tmp */
  silc_mp_mul(dst, dst, &key->q);
  silc_mp_add(dst, dst, &tmp);

  silc_mp_uninit(&tmp);

  return TRUE;
}
